### ISSUE TITLE
Bug in NStepLSTM when only hx output is used

### ISSUE THREAD

**Issue by fabiencro** [pfnet/chainer#2505]
Hi,

I think I found a subtle bug in the n_step_lstm links and functions. It manifests itself randomly, so it is not so easy to reproduce. It might be a problem with my system, but I checked with two machines (both with ubuntu 16.04 and Cudnn 5.1 and Pascal GPUs) and had this bug in both.

This bug seems to happen when we do a backward on a computation graph that did not use the "xs" variables returned by the link, but only the "hx" component. In such case, the computed gradients will *sometimes* be full of NaN.

This might not be totally clear, so here is a piece of code that can produce the bug:

```
import chainer
import chainer.links as L
import chainer.functions as F
from chainer import Variable
import numpy as np
import argparse

# command line parsing
parser = argparse.ArgumentParser(formatter_class=argparse.ArgumentDefaultsHelpFormatter)
parser.add_argument("--gpu", type = int)
parser.add_argument("--length", type = int, default = 12)
parser.add_argument("--use_workaround", default = False, action = "store_true")
args = parser.parse_args()

# Setting up gpu
if args.gpu is not None:
    chainer.cuda.Device(args.gpu).use()
    import cupy
    xp = cupy
else:
    xp = np

# Turn debug mode on
chainer.set_debug(True)

# Create some links for testing (model will be NStepLSTM -> Linear -> Softmax)
nstep = L.NStepLSTM(1, 32, 64, dropout = 0.5)
lin = L.Linear(64, 10)

if args.gpu is not None:
    nstep = nstep.to_gpu(args.gpu)
    lin = lin.to_gpu(args.gpu)
    
# Create some input data and target data
xa = np.random.randn(args.length, 32).astype(np.float32)
tgt = np.array([3], dtype = np.int32)

if args.gpu is not None:
    xa = cupy.array(xa)
    tgt = cupy.array(tgt)
    
xv = Variable(xa)
tgt_v = Variable(tgt)
  
hx, cx, xs = nstep(None, None, (xv,))
loss = F.softmax_cross_entropy(lin(hx), tgt_v)

# Using this workaround remove the bug
if args.use_workaround:
    loss += 0*(F.sum(xs[0]))

print loss.data
loss.backward()  # will sometimes generate NaN
```

This script has 3 options: --gpu, --length and --use_workaround.

If the script is run without the gpu option (ie. we are using numpy implementation), it always work fine.
However, if we are using the --gpu #num_gpu option, there is a chance the script will exit with an error because the debug mode of chainer has detected NaN in the computed gradients. This will not always happen, though. One can increase the --length parameter to increase the chance of the bug happening. I have seen it happen with length values as low as 10. But sometimes I have to use --length 10000 or more to have the bug.

If then, one use the --use_workaround option, the bug disappear. (the use_workaround option makes it so that the xs variables will be part of the computation graph).

I tried to hunt for the cause of the bug. My current hypothesis is that, in the backward code of n_step_lstm, if the gradient of the xs outputs is None, some zero-filled cupy arrays are created to pass to the cudnn backward function. It might be that in such case, the backward function exit before the asynchronous gpu computation is finished, thus releasing the memory of the zero-filled arrays when the cudnn computation is still going on.

At this point, I do not understand cupy and cudnn enough check further. So hopefully somebody can try and reproduce this bug and then solve it.

**Comment by unnonouno** [comment]
Thank you for reporting a problem. I'll check it.

**Comment by unnonouno** [comment]
I can reproduce the error with cuDNN v5 and v5.1. But, with cuDNN v6 the error doesn't happen. cuDNN v6 the latest version is supported from Chainer v 1.23.0 that is released today. Please try the newest version.

**Comment by fabiencro** [comment]
Unfortunately, I am using a shared computational infrastructure, so I cannot upgrade cuDNN at will. For now I am using the workaround of adding 0*F.sum(xs) to the final loss, but it is inefficient.

It is nice to know this is solved with cuDNN v6, though. But have you understood what is causing the problem? Do you think my guess that the cupy arrays (that are created in backward when the out_grad is None) are released too early might be correct?

**Comment by unnonouno** [comment]
Try to use cudnnenv which install cudnn in your home directory and switch its version.  https://pypi.python.org/pypi/cudnnenv
I think it is a bug of cudnn simply. As cudnn has many bugs, i recommend to use latest version.

**Comment by fabiencro** [comment]
OK, I did not know about cudnnenv. Thank you for the tip :-)
